### PR TITLE
Refactor the lowering of remat

### PR DESCRIPTION
**Pull request by gnecula**
Replace the XLA-specific lowering with the same logic written
with LAX primitives. This will allow reusing the logic for
different lowerings, e.g., jax2tf or MLIR.

### CODE REVIEW

**Comment by hawkinsp** [comment]
There's another copy of this logic here: https://github.com/google/jax/blob/c1ce85203c65cd816e98422122b25320eb33539c/jax/interpreters/mlir.py#L929
which you should probably change to use the new version. It will look very similar to the copy in `xla.py`.

**Comment by gnecula** [comment]
> There's another copy of this logic here:
> 
> https://github.com/google/jax/blob/c1ce85203c65cd816e98422122b25320eb33539c/jax/interpreters/mlir.py#L929
> 
> 
> which you should probably change to use the new version. It will look very similar to the copy in `xla.py`.

I saw that, but I could not yet figure out how to call the xla lowering. Is there an example? It seemed to me that all the primitives where this would be needed are done by way of fallback, instead of just interpreting the LAX function (a la `lower_fun`).

**Comment by gnecula** [comment]
> You shouldn't be calling the XLA lowering, you should be using `mlir.lower_fun` to lower the traceable JAX version to MLIR directly.

Done.

File: jax/_src/lax/control_flow.py [hunk 1]
```diff
@@ -2924,3 +2925,92 @@ def _cumulative_jvp_rule(primals, tangents, *, axis: int, reverse: bool,
...
+
+
+def _dummy_remat_result(aval: core.AbstractValue):
+  """A result that will be discarded"""
+  if aval is core.AbstractToken:
+    return lax.create_token()
+  elif aval is core.abstract_unit:
+    return core.unit
+  else:
+    return lax.broadcast(np.array(0, dtype=aval.dtype), aval.shape)  # type: ignore
```
Comment (hawkinsp) [inline jax/_src/lax/control_flow.py:2937]: lax.full_like ?

**Review by hawkinsp — APPROVED** [review]
Verify this works with MLIR enabled. I'm about to enable internal CI tests for this.

File: jax/_src/lax/control_flow.py (hunk 1 above)
Comment (gnecula) [reply jax/_src/lax/control_flow.py:2937]: For `lax.full_like` I need a value x to use the shape or dtype for. I could write `lax.full_like(None, 0, dtype=aval.dtype, shape=aval.shape)` but I think it is less clear that this is just a broadcast.

**Pull request merged**